datastream-kafka/src/main/java/com/linkedin/datastream/kafka/KafkaTransportProvider.java

```diff
@@ -136,6 +143,38 @@ private int getSourcePartitionFromEvent(BrooklinEnvelope event) {
         event.getMetadata().getOrDefault(BrooklinEnvelopeMetadataConstants.SOURCE_PARTITION, "-1"));
   }
 
+  @Override
+  public void broadcast(String destinationUri, DatastreamProducerRecord record, SendBroadcastCallback onSendComplete) {
+    Validate.isTrue(record.isBroadcastRecord(), "Trying to broadcast a non-broadcast type record.");
+
+    Properties adminClientProps = new Properties();
+    adminClientProps.putAll(_transportProviderProperties);
+    AdminClient adminClient = AdminClient.create(adminClientProps);
+    String topicName = KafkaTransportProviderUtils.getTopicName(destinationUri);
+    int partitionCount = -1;
+    List<Pair<DatastreamRecordMetadata, Exception>> listMetadataExceptionPair = new ArrayList<>();
+
+    try {
+      // Extract partition count from the topic using AdminClient
+      partitionCount =
+          adminClient.describeTopics(Collections.singleton(topicName)).all().get().get(topicName).partitions().size();
+    } catch (ExecutionException | InterruptedException ex) {
+      LOG.error("Failed to parse partition count for topic {} required for broadcast", topicName);
+      listMetadataExceptionPair.add(new Pair<>(null, ex));
+      onSendComplete.onCompletion(listMetadataExceptionPair, partitionCount);
+      return;
+    }
+
+    LOG.debug("Broadcasting record {} to all {} partitions of destination {}", record, partitionCount, destinationUri);
+    for (int i = 0; i < partitionCount; i++) {
+      record.setPartition(i);
+      send(destinationUri, record, ((metadata, exception) -> {
+        listMetadataExceptionPair.add(new Pair<>(metadata, exception));
+      }));
+    }
+    onSendComplete.onCompletion(listMetadataExceptionPair, partitionCount);
```

Comment:
This should be inside the send() code, otherwise the oncompletion will get called before send completes.